plone/app/vocabularies/principals.py

```diff
@@ -0,0 +1,200 @@
+
+
+@implementer(ISlicableVocabulary)
+class PrincipalsVocabulary(SimpleVocabulary):
+    """Vocabulary dealing with users/ groups (or in theory any other principal)
+    """
+
+    @property
+    def principal_source(self):
+        return self._principal_source
+
+    @principal_source.setter
+    def principal_source(self, value):
+        self._principal_source = value
+
+    @property
+    def _acl_users(self):
+        aclu = getattr(self, '_aclu', None)
+        if not aclu:
+            aclu = self._aclu = _get_acl_users()
+        return aclu
+
+    def _get_from_source(self, value, default=None):
+        """Helper to get a user or group from users folder.
+        """
+        if SOURCES[self._principal_source]['prefix']:
+            if ':' not in value:
+                return default
+            principal_type, principal_id = value.split(':', 2)
+        else:
+            principal_type = self._principal_source
+            principal_id = value
+        getter = getattr(self._acl_users, GETTER[principal_type])
+        return getter(principal_id, default)
+
+    def __contains__(self, value):
+        """Checks if the principal exists in current subset or in PAS.
+        """
+        result = super(PrincipalsVocabulary, self).__contains__(value)
+        return result or bool(self._get_from_source(value))
+
+    def getTerm(self, value):
+        """Checks also for values not in the current subset.
+        This allows to lookup already saved values.
+        """
+        try:
+            return super(PrincipalsVocabulary, self).getTerm(value)
+        except LookupError:
+            user = self._get_from_source(value)
+            if user is None:
+                raise
+            if user.isGroup():
+                title = user.getProperty('title', value) or value
+            else:
+                title = user.getProperty('fullname', value) or value
+            if SOURCES[self._principal_source]['prefix']:
+                token = value.replace(':', '__', 1)
+            else:
+                token = value
+            return self.__class__.createTerm(value, token, title)
+
+    def __getitem__(self, start, stop=None):
+        """Sliceable"""
+        if isinstance(start, slice):
+            slice_inst = start
+            start = slice_inst.start
+            stop = slice_inst.stop
+        elif not stop:
+            return self._terms[start]
+
+        # sliced up
+        return self._terms[start:stop]
+
+
+class BaseFactory(object):
+    """Factory creating a PrincipalsVocabulary
+    """
+
+    source = None
+
+    def should_search(self, query):
+        """ Test if we should search for users
+        """
+        if query:
+            return True
+        registry = getUtility(IRegistry)
+        return not [
+            x
+            for x in filter(
+                registry.get,
+                [cfg['many'] for cfg in SOURCES[self.source]['searches']],
+            )
+        ]
```

Comment:
A more performant way:
```
for x in filter(...):
    return True
return False
```

Reply:
Well, the whole list contains at maximum 2 values.

Reply:
Right :) Anyway I already approved the PR, it was just a suggestion to maybe make it more readable

Reply:
Thanks! Please wait with merging for @pbauer testing this if it still speeds up 200x in his customers project.